drivers/dac/dac_dacx0501.c

```diff
@@ -0,0 +1,197 @@
+	REF_EXTERNAL, /* External pin voltage reference. */
+};
+
+/* Specifies the reference voltage multiplier. */
+enum voltage_mult {
+	VM_MUL2, /* Multiplies by 2. */
+	VM_MUL1, /* Multiplies by 1. */
+	VM_DIV2, /* Multiplies by 0.5 */
+};
+
+struct dacx0501_config {
+	struct i2c_dt_spec i2c_spec;
+	enum ref_voltage_source ref_source;
+	enum voltage_mult voltage_mult;
+};
+
+struct dacx0501_data {
+	/* Device can only be configured once. */
+	bool configured;
+	/* Number of bits in the DAC register: Either 12, 14 or 16. */
+	uint8_t resolution;
+};
+
+static int dacx0501_reg_read(const struct device *dev, const uint8_t addr, uint16_t *data)
+{
+	const struct dacx0501_config *config = dev->config;
+	uint8_t raw_data[2];
+	int status;
+
+	status = i2c_write_read_dt(&config->i2c_spec, &addr, sizeof(addr), raw_data,
+				   sizeof(raw_data));
+	if (status) {
+		return status;
+	}
+
+	/* DAC is big endian. */
+	*data = sys_get_be16(raw_data);
+	return 0;
+}
+
+static int dacx0501_reg_write(const struct device *dev, uint8_t addr, uint16_t data)
+{
+	const struct dacx0501_config *config = dev->config;
+	uint8_t write_cmd[3] = {addr};
+
+	/* DAC is big endian. */
+	sys_put_be16(data, write_cmd + 1);
+
+	return i2c_write_dt(&config->i2c_spec, write_cmd, sizeof(write_cmd));
+}
+
+static int dacx0501_channel_setup(const struct device *dev,
+				  const struct dac_channel_cfg *channel_cfg)
+{
+	const struct dacx0501_config *config = dev->config;
+	struct dacx0501_data *data = dev->data;
+	uint16_t device_id;
+	int status;
+
+	/* DACx0501 series only has a single output channel. */
+	if (channel_cfg->channel_id != 0) {
+		LOG_ERR("Unsupported channel %d", channel_cfg->channel_id);
+		return -ENOTSUP;
+	}
+
+	if (data->configured) {
+		LOG_ERR("Channel already configured");
+		return -EINVAL;
+	}
+
+	if (!i2c_is_ready_dt(&config->i2c_spec)) {
+		LOG_ERR("I2C bus %s not ready", config->i2c_spec.bus->name);
+		return -ENODEV;
+	}
+
+	status = dacx0501_reg_read(dev, DACX0501_REG_DEVICE_ID, &device_id);
+	if (status) {
+		LOG_ERR("read DEVICE_ID register failed");
+		return status;
+	}
+
+	/* See DEVICE_ID register RES field in the data sheet. */
+	data->resolution = 16 - 2 * FIELD_GET(DACX0501_MASK_DEVICE_ID_RES, device_id);
+
+	if (channel_cfg->resolution != data->resolution) {
+		LOG_ERR("Unsupported resolution %d. Actual: %d", channel_cfg->resolution,
+			data->resolution);
+		return -ENOTSUP;
+	}
+
+	status = dacx0501_reg_write(
+		dev, DACX0501_REG_CONFIG,
+		FIELD_PREP(DACX0501_MASK_CONFIG_REF_PWDWN, config->ref_source == REF_EXTERNAL));
+	if (status) {
+		LOG_ERR("write CONFIG register failed");
+		return status;
+	}
+
+	status = dacx0501_reg_write(
+		dev, DACX0501_REG_GAIN,
+		FIELD_PREP(DACX0501_MASK_GAIN_REFDIV_EN, config->voltage_mult == VM_DIV2) |
+			FIELD_PREP(DACX0501_MASK_GAIN_BUFF_GAIN, config->voltage_mult == VM_MUL2));
+	if (status) {
+		LOG_ERR("GAIN Register update failed");
+		return status;
+	}
```

Comment:
I think this code should all go into the driver init function, similar to the `dacx3608` driver. It is more a device-level check / initialization and would not need to be called for each channel (if we had more than one channel).

Reply:
Done

Reply:
Doesn't seem to be done.